File: Olympiad/COCI/official/2007/contest5_solutions/tenis.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 5 - Task TENIS
  Programming language Pascal
*)

program tenis;

function provjeri_jedan_set(g1, g2, koji_set : integer;
                            ime1, ime2       : string;
                            var pob          : array of integer):boolean;
var
   t : integer;
   p : string;
begin
   if g1 = g2 then begin
      provjeri_jedan_set := false;
      exit;
   end;

   if g1 < g2 then begin
      t := g1; g1 := g2; g2 := t;
      p := ime1; ime1 := ime2; ime2 := p;
      pob[1] := pob[1] + 1;
   end else
      pob[0] := pob[0] + 1;

   if ime2 = 'federer' then
      provjeri_jedan_set := false
   else if g1 < 6 then
      provjeri_jedan_set := false
   else if g1 = 6 then
      provjeri_jedan_set := (g2 <= 4)
   else if (g1 = 7) and (koji_set < 3) then
      provjeri_jedan_set := (g2 = 5) or (g2 = 6)
   else if (g1 >= 7) and (koji_set = 3) then
      provjeri_jedan_set := (g1 = g2+2)
   else
      provjeri_jedan_set := false;
end;

var
   red, ime1, ime2         : string;
   n, i, poc, mec          : integer;
   koji_set, g1, g2, smece : integer;
   ok                      : boolean;
   pob                     : array[1..2] of integer;

begin
   readln(red);
   i := pos(' ', red);
   ime1 := copy(red, 1, i-1);
   ime2 := copy(red, i+1, length(red)-i);
   readln(n);
   for mec:=1 to n do begin
      readln(red);
      red := red + ' ';

      pob[1] := 0;
      pob[2] := 0;
      koji_set := 1;
      poc := 1;
      ok := true;

      for i:=1 to length(red) do begin
         if red[i]=':' then begin
            val(copy(red, poc, i-poc), g1, smece);
            poc := i+1;
         end else if red[i]=' ' then begin
            val(copy(red, poc, i-poc), g2, smece);
            poc := i+1;
            ok := ok and (pob[1] < 2) and (pob[2] < 2);
            ok := ok and provjeri_jedan_set(g1, g2, koji_set, ime1, ime2, pob);
            koji_set := koji_set + 1;
         end;
      end;

      ok := ok and ((pob[1] = 2) or (pob[2] = 2));
      if ok then
         writeln('da')
      else
         writeln('ne');
   end;
end.

File: Olympiad/COCI/official/2007/contest5_solutions/tenis.c

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 5 - Task TENIS
  Programming language C
*/

#include <stdio.h>
#include <string.h>

int provjeri_jedan_set( int g1, int g2, int koji_set, const char *ime1, const char *ime2, int pob[2] )
{
   if ( g1 == g2 ) return 0;

   if ( g1 < g2 ) {
      int t; const char *p;

      t = g1; g1 = g2; g2 = t;
      p = ime1; ime1 = ime2; ime2 = p;
      ++pob[1];
   }
   else
      ++pob[0];

   if ( strcmp( ime2, "federer" ) == 0 ) return 0;

   if ( g1 < 6 ) return 0;
   if ( g1 == 6 ) return g2 <= 4;
   if ( g1 == 7 && koji_set < 3 ) return g2 == 5 || g2 == 6;
   if ( g1 >= 7 && koji_set == 3 ) return g1 == g2+2;
   return 0;
}

int main()
{
   char ime1[21], ime2[21];
   int n, i;
   scanf( "%s%s", ime1, ime2 );
   scanf( "%d\n", &n );

   for ( i=0; i<n; ++i ) {
      int ok = 1;
      char red[100];
      char *p = red;
      int pob[2] = { 0, 0 };
      int koji_set, g1, g2, ucitano;

      fgets( red, sizeof red, stdin );

      for ( koji_set=1; sscanf( p, "%d:%d%n", &g1, &g2, &ucitano ) == 2; ++koji_set ) {
         p += ucitano;
         ok &= pob[0] < 2 && pob[1] < 2;
         ok &= provjeri_jedan_set( g1, g2, koji_set, ime1, ime2, pob );
      }
      ok &= pob[0] == 2 || pob[1] == 2;

      puts( ok ? "da" : "ne" );
   }

   return 0;
}
